packages/@aws-cdk/aws-s3/test/auto-delete-objects.test.ts

```diff
@@ -0,0 +1,318 @@
+});
+
+test('when only one bucket has autoDeleteObjects enabled, only that bucket gets assigned a custom resource', () => {
+  // GIVEN
+  const app = new cdk.App();
+  const stack = new cdk.Stack(app);
+
+  // WHEN
+  new s3.Bucket(stack, 'MyBucketOne', {
+    removalPolicy: cdk.RemovalPolicy.DESTROY,
+    autoDeleteObjects: true,
+  });
+  new s3.Bucket(stack, 'MyBucketTwo', {
+    removalPolicy: cdk.RemovalPolicy.DESTROY,
+    autoDeleteObjects: false,
+  });
+
+  // THEN
+  const template = app.synth().getStackArtifact(stack.artifactId).template;
+  const resourceTypes = Object.values(template.Resources).map((r: any) => r.Type).sort();
+
+  expect(resourceTypes).toStrictEqual([
+    // custom resource provider resources
+    'AWS::IAM::Role',
+    'AWS::Lambda::Function',
+
+    // buckets
+    'AWS::S3::Bucket',
+    'AWS::S3::Bucket',
+
+    // auto delete object resource
+    'Custom::AutoDeleteObjects',
+  ]);
+
+  // custom resource for MyBucket1 is present
+  expect(stack).toHaveResource('Custom::AutoDeleteObjects',
+    { BucketName: { Ref: 'MyBucketOneA6BE54C9' } });
+
+  // custom resource for MyBucket2 is not present
+  expect(stack).not.toHaveResource('Custom::AutoDeleteObjects',
+    { BucketName: { Ref: 'MyBucketTwoC7437026' } });
+});
+
+test('iam policy is created', () => {
+  // GIVEN
+  const stack = new cdk.Stack();
+
+  // WHEN
+  new s3.Bucket(stack, 'MyBucket', {
+    removalPolicy: cdk.RemovalPolicy.DESTROY,
+    autoDeleteObjects: true,
+  });
+
+  // THEN
+  expect(stack).toHaveResource('AWS::IAM::Role', {
+    Policies: [
+      {
+        PolicyName: 'Inline',
+        PolicyDocument: {
+          Version: '2012-10-17',
+          Statement: [
+            {
+              Effect: 'Allow',
+              Resource: '*',
+              Action: [
+                's3:GetObject*',
+                's3:GetBucket*',
+                's3:List*',
+                's3:DeleteObject*',
+              ],
+            },
+          ],
+        },
+      },
+    ],
+  });
+});
+
+test('throws if autoDeleteObjects is enabled but if removalPolicy is not specified', () => {
+  // GIVEN
+  const stack = new cdk.Stack();
+
+  // WHEN
+  // THEN
+  expect(() => new s3.Bucket(stack, 'MyBucket', { autoDeleteObjects: true })).toThrowError(/removal policy/);
+});
+
+describe('custom resource handler', () => {
+
+  const mockS3Client = {
+    listObjectVersions: jest.fn().mockReturnThis(),
+    deleteObjects: jest.fn().mockReturnThis(),
+    promise: jest.fn(),
+  };
+
+  jest.mock('aws-sdk', () => {
+    return { S3: jest.fn(() => mockS3Client) };
+  });
+
+  beforeEach(() => {
+    // must reset implementations otherwise it's possible for
+    // mockResolvedValue values to leak between tests
+    mockS3Client.promise.mockReset();
+    mockS3Client.listObjectVersions.mockReset();
+    mockS3Client.deleteObjects.mockReset();
+
+    mockS3Client.listObjectVersions.mockReturnThis();
+    mockS3Client.deleteObjects.mockReturnThis();
+  });
```

Comment:
So typically in jest world, the mocking is done in `beforeEach` and cleanup in `afterEach`. Also, cleanup can probably be less verbose as well

```ts
beforeEach(() => {
	mockS3Client.listObjectVersions.mockReturnThis();
    mockS3Client.deleteObjects.mockReturnThis();
});

afterEach(() => {
	jest.clearAllMocks(); // or jest.resetAllMocks()
});
```

Reply:
Good point :)